iree/runtime/setup.py

```diff
@@ -0,0 +1,379 @@
+
+
+def prepare_installation():
+  subprocess.check_call(["cmake", "--version"])
+  version_py_content = generate_version_py()
+  print(f"Generating version.py:\n{version_py_content}", file=sys.stderr)
+
+  if not IS_CONFIGURED:
+    # Build from source tree.
+    os.makedirs(IREE_BINARY_DIR, exist_ok=True)
+    maybe_nuke_cmake_cache()
+    print(f"CMake build dir: {IREE_BINARY_DIR}", file=sys.stderr)
+    print(f"CMake install dir: {CMAKE_INSTALL_DIR_ABS}", file=sys.stderr)
+    cfg = "Release"
+    cmake_args = [
+        "-GNinja",
+        "--log-level=VERBOSE",
+        "-DIREE_BUILD_PYTHON_BINDINGS=ON",
+        "-DIREE_BUILD_COMPILER=OFF",
+        "-DIREE_BUILD_SAMPLES=OFF",
+        "-DIREE_BUILD_TESTS=OFF",
+        "-DPython3_EXECUTABLE={}".format(sys.executable),
+        "-DCMAKE_BUILD_TYPE={}".format(cfg),
+        get_env_cmake_option("IREE_HAL_DRIVER_CUDA"),
+        get_env_cmake_option("IREE_ENABLE_RUNTIME_TRACING"),
+        get_env_cmake_option("IREE_BUILD_TRACY"),
+    ]
+
+    # Only do a from-scratch configure if not already configured.
+    cmake_cache_file = os.path.join(IREE_BINARY_DIR, "CMakeCache.txt")
+    if not os.path.exists(cmake_cache_file):
+      print(f"Configuring with: {cmake_args}", file=sys.stderr)
+      subprocess.check_call(["cmake", IREE_SOURCE_DIR] + cmake_args,
+                            cwd=IREE_BINARY_DIR)
+    else:
+      print(f"Not re-configuring (already configured)", file=sys.stderr)
+
+    # Build.
+    subprocess.check_call(
+        ["cmake", "--build", ".", "--target", "iree/runtime/python/all"],
+        cwd=IREE_BINARY_DIR)
+    print("Build complete.", file=sys.stderr)
+
+  # Install the directory we care about.
+  install_subdirectory = os.path.join(IREE_BINARY_DIR, "iree", "runtime",
+                                      "python", "iree", "runtime")
+  install_args = [
+      "-DCMAKE_INSTALL_DO_STRIP=ON",
+      f"-DCMAKE_INSTALL_PREFIX={CMAKE_INSTALL_DIR_ABS}/",
+      "-P",
+      os.path.join(install_subdirectory, "cmake_install.cmake"),
+  ]
+  print(f"Installing with: {install_args}", file=sys.stderr)
+  subprocess.check_call(["cmake"] + install_args, cwd=install_subdirectory)
+
+  # Write version.py directly into install dir.
+  version_py_file = os.path.join(CMAKE_INSTALL_DIR_ABS, "python_packages",
+                                 "iree_runtime", "iree", "runtime",
+                                 "version.py")
+  os.makedirs(os.path.dirname(version_py_file), exist_ok=True)
+  with open(version_py_file, "wt") as f:
+    f.write(version_py_content)
+
+  print(f"Installation prepared: {CMAKE_INSTALL_DIR_ABS}", file=sys.stderr)
+
+
+class CMakeBuildPy(_build_py):
+
+  def run(self):
+    # It is critical that the target directory contain all built extensions,
+    # or else setuptools will helpfully compile an empty binary for us
+    # (this is the **worst** possible thing it could do). We just copy
+    # everything. What's another hundred megs between friends?
```

Comment:
And this is done without checking if there already? I'm guessing as this isn't that heavy to optimize avoiding copy/too brittle to try

Reply:
Well, it will check if it is in the target directory -- not whether it is in the source directory. In general, this is the only reliable way I have found to disarm all versions of setuptools from attempting to invoke the C compiler themselves. It is a real mess.